src/main/scala/com/wavesplatform/state2/reader/StateReader.scala

```diff
@@ -0,0 +1,124 @@
+package com.wavesplatform.state2.reader
+
+import com.google.common.base.Charsets
+import com.wavesplatform.state2._
+import scorex.account.{Account, AccountOrAlias, Alias}
+import scorex.crypto.hash.FastCryptographicHash
+import scorex.transaction.ValidationError.{AliasNotExists, TransactionValidationError}
+import scorex.transaction._
+import scorex.transaction.assets.IssueTransaction
+import scorex.transaction.assets.exchange.{ExchangeTransaction, Order}
+import scorex.transaction.lease.LeaseTransaction
+import scorex.transaction.state.database.state.AddressString
+
+import scala.reflect.ClassTag
+import scala.util.Right
+
+trait StateReader {
+
+  def accountPortfolios: Map[Account, Portfolio]
+
+  def transactionInfo(id: ByteArray): Option[(Int, Transaction)]
+
+  def accountPortfolio(a: Account): Portfolio
+
+  def assetInfo(id: ByteArray): Option[AssetInfo]
+
+  def height: Int
+
+  def accountTransactionIds(a: Account): Seq[ByteArray]
+
+  def nonEmptyAccounts: Seq[Account]
+
+  def effectiveBalanceAtHeightWithConfirmations(acc: Account, height: Int, confs: Int): Long
+
+  def paymentTransactionIdByHash(hash: ByteArray): Option[ByteArray]
+
+  def maxPaymentTransactionTimestampInPreviousBlocks(a: Account): Option[Long]
+
+  def aliasesOfAddress(a: Account): Seq[Alias]
+
+  def resolveAlias(a: Alias): Option[Account]
+
+  def findPreviousExchangeTxs(orderId: EqByteArray): Set[ExchangeTransaction]
+
+  def isLeaseActive(leaseTx: LeaseTransaction): Boolean
+}
+
+object StateReader {
+
+  implicit class StateReaderExt(s: StateReader) {
+    def assetDistribution(assetId: ByteArray): Map[Account, Long] =
+      s.nonEmptyAccounts
+        .flatMap(acc => s.accountPortfolio(acc).assets.get(assetId).map(acc -> _))
+        .toMap
+
+    def findTransaction[T <: Transaction](signature: Array[Byte])(implicit ct: ClassTag[T]): Option[T]
+    = s.transactionInfo(EqByteArray(signature)).map(_._2)
+      .flatMap(tx => {
+        if (ct.runtimeClass.isAssignableFrom(tx.getClass))
+          Some(tx.asInstanceOf[T])
+        else None
+      })
+
+    def resolveAliasEi[T <: Transaction](aoa: AccountOrAlias): Either[StateValidationError, Account] = {
+      aoa match {
+        case a: Account => Right(a)
+        case a: Alias => s.resolveAlias(a) match {
+          case None => Left(AliasNotExists(a))
+          case Some(acc) => Right(acc)
+        }
+      }
+    }
+
+    def findPreviousExchangeTxs(order: Order): Set[ExchangeTransaction] =
+      s.findPreviousExchangeTxs(EqByteArray(order.id))
+
+    def included(signature: Array[Byte]): Option[Int] = s.transactionInfo(EqByteArray(signature)).map(_._1)
```

Comment:
This function is part of `implicit class StateReaderExt(s: StateReader)`.
This prevented compilation errors at places where old `State` with the exact same signature was used. Now it's `StateReader` with this extension. Most of these extension methods are not needed at all and can be inlined, I haven't done this to prevent merge conflicts in case someone decides to rely on these in `master` branch.